libflux/src/core/lib.rs

```diff
@@ -0,0 +1,481 @@
+#![cfg_attr(feature = "strict", deny(warnings, missing_docs))]
+#![allow(clippy::unknown_clippy_lints)]
+//! The flux crate handles the parsing and semantic analysis of flux source
+//! code.
+extern crate chrono;
+#[macro_use]
+extern crate serde_derive;
+extern crate serde_aux;
+
+pub mod ast;
+pub mod formatter;
+pub mod parser;
+pub mod scanner;
+pub mod semantic;
+
+use std::error;
+use std::ffi::*;
+use std::fmt;
+use std::os::raw::c_char;
+
+use parser::Parser;
+
+pub use ast::DEFAULT_PACKAGE_NAME;
+
+/// An error handle designed to allow passing `Error` instances to library
+/// consumers across language boundaries.
+pub struct ErrorHandle {
+    /// A heap-allocated `Error`
+    pub err: Box<dyn error::Error>,
+}
+
+/// An error that can occur due to problems in ast generation or semantic
+/// analysis.
+#[derive(Debug, Clone)]
+pub struct Error {
+    msg: String,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str(&self.msg)
+    }
+}
+
+impl error::Error for Error {
+    fn source(&self) -> Option<&(dyn error::Error + 'static)> {
+        None
+    }
+}
+
+impl From<String> for Error {
+    fn from(msg: String) -> Self {
+        Error { msg }
+    }
+}
+
+impl From<&str> for Error {
+    fn from(msg: &str) -> Self {
+        Error {
+            msg: String::from(msg),
+        }
+    }
+}
+
+impl From<semantic::nodes::Error> for Error {
+    fn from(sn_err: semantic::nodes::Error) -> Self {
+        Error { msg: sn_err.msg }
+    }
+}
+
+impl From<semantic::check::Error> for Error {
+    fn from(err: semantic::check::Error) -> Self {
+        Error {
+            msg: format!("{}", err),
+        }
+    }
+}
+
+/// Frees a previously allocated error.
+///
+/// Note: we use the pattern described here: https://doc.rust-lang.org/std/boxed/index.html#memory-layout
+/// wherein a pointer where ownership is being transferred is modeled as a Box, and if it could be
+/// null, then it's wrapped in an Option.
+#[no_mangle]
+pub extern "C" fn flux_free_error(_err: Option<Box<ErrorHandle>>) {}
```

Comment:
I'd like to see all of these methods moved into `flux/lib.rs` so that the C API implementation all lives in that single file.

Reply:
Will do.

Reply:
@nathanielc would you mind to have a look at the new commit?